vro:pull command for legacy archetype fails when workflow folder path name contains character &

### Description

When running a vro:pull command on legacy workflow archetype, if those workflows paths name contains the & character, the pull command will faill with the log below


### Steps to Reproduce

1. Have a vro package containing workflows which paths names (folder name) include the &
2. Try a vro:pull command of that package
3. See the maven vro:pull command fails

**Preconditions:** Have the "&" character in the workflow folder path name

**Expected behavior:** vRBT to handle this character as any other character and succeed

**Actual behavior:** It fails when running the vro:pull command

**Reproduces how often:** 100%

**Component/s:** maven/plugins/vropkg

**Affects Build/s:** 2.32.0

### Environment

#### Client

-   Build Tools for VMware Aria Version: 2.32.0
-   Visual Studio Code Version:  1.70
-   OS Version: W10

#### Server

-   vRealize Automation Version: 8.11.1
-   vRealize Orchestrator Version: 8.11.1
-   vRealize Operations Version: n/a
-   vRealize Log Insight Version: n/a

### Failure Logs
```
[INFO] C:\Users\username\projects\windows_vra_catalog\legacy\workflows>node "C:\Users\username\projects\windows_vra_catalog\legacy\workflows\node_modules\@vmware-pscoe\vropkg\bin\\..\dist\cli.js" --in flat --srcPath C:\Users\LADM_T~1\AppData\Local\Temp\2\vro-xml-pull1906541284324290868\com.company.vrbt.winvra.legacy-1.0.0-SNAPSHOT.package --out tree --destPath C:\Users\username\projects\windows_vra_catalog\legacy\workflows --privateKeyPEM C:\.keystore\private_key.pem --certificatesPEM C:\.keystore\cert.pem --keyPass VMware1! --version 1.0.0-SNAPSHOT --packaging package --artifactId legacy --description "Build Tools for VMware Aria provides development and release management tools for implementing automation solutions                 based on the VMware Aria Suite and VMware Cloud Director. The solution enables Virtual Infrastructure Administrators and                  Automation Developers to use standard DevOps practices for managing and deploying content." --groupId com.company.vrbt.winvra
[INFO] info: Extracting package C:\Users\LADM_T~1\AppData\Local\Temp\2\vro-xml-pull1906541284324290868\com.company.vrbt.winvra.legacy-1.0.0-SNAPSHOT.package to "C:\Users\username\projects\windows_vra_catalog\legacy\workflows" folder...
[ERROR] C:\Users\username\projects\windows_vra_catalog\legacy\workflows\node_modules\@vmware-pscoe\vropkg\dist\serialize\tree.js:85
[ERROR]         : element.categoryPath.map(c => c.replace(/\./g, "/."));
[ERROR]                                           ^
[ERROR]
[ERROR] TypeError: Cannot read properties of undefined (reading 'replace')
[ERROR]     at C:\Users\username\projects\windows_vra_catalog\legacy\workflows\node_modules\@vmware-pscoe\vropkg\dist\serialize\tree.js:85:43
[ERROR]     at Array.map (<anonymous>)
[ERROR]     at C:\Users\username\projects\windows_vra_catalog\legacy\workflows\node_modules\@vmware-pscoe\vropkg\dist\serialize\tree.js:85:32
[ERROR]     at Generator.next (<anonymous>)
[ERROR]     at C:\Users\username\projects\windows_vra_catalog\legacy\workflows\node_modules\@vmware-pscoe\vropkg\dist\serialize\tree.js:8:71
[ERROR]     at new Promise (<anonymous>)
[ERROR]     at __awaiter (C:\Users\username\projects\windows_vra_catalog\legacy\workflows\node_modules\@vmware-pscoe\vropkg\dist\serialize\tree.js:4:12)
[ERROR]     at serializeTreeElement (C:\Users\username\projects\windows_vra_catalog\legacy\workflows\node_modules\@vmware-pscoe\vropkg\dist\serialize\tree.js:73:52)
[ERROR]     at C:\Users\username\projects\windows_vra_catalog\legacy\workflows\node_modules\@vmware-pscoe\vropkg\dist\serialize\tree.js:118:37
[ERROR]     at Array.forEach (<anonymous>)
[ERROR]     at Object.serializeTree (C:\Users\username\projects\windows_vra_catalog\legacy\workflows\node_modules\@vmware-pscoe\vropkg\dist\serialize\tree.js:118:18)
[ERROR]     at C:\Users\username\projects\windows_vra_catalog\legacy\workflows\node_modules\@vmware-pscoe\vropkg\dist\cli.js:124:24
[ERROR]     at Generator.next (<anonymous>)
[ERROR]     at fulfilled (C:\Users\username\projects\windows_vra_catalog\legacy\workflows\node_modules\@vmware-pscoe\vropkg\dist\cli.js:5:58)
[INFO] Running vropkg... finished
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for legacy-all 1.0.0-SNAPSHOT:
[INFO] 
[INFO] legacy.actions ..................................... SUCCESS [01:36 min]
[INFO] legacy ............................................. FAILURE [01:21 min]
[INFO] legacy-all ......................................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  03:36 min
[INFO] Finished at: 2023-09-05T15:27:42+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal com.vmware.pscoe.maven.plugins:o11n-xml-package-maven-plugin:2.35.0:pull (default-cli) on project legacy: Running vropkg... failed with code 1 -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR]
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <args> -rf :legacy
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -3,6 +3,7 @@
 
 ### Fixes
 * [vro-types] Fix SSHSession 'error' and 'state' types
+* [typescript] 165 / vro:pull command for legacy archetype fails when workflow folder path name contains character '&'.
 
 ## v2.37.0 - 26 Jan 2024
 

--- a/docs/versions/latest/Release.md
+++ b/docs/versions/latest/Release.md
@@ -36,12 +36,24 @@
 [//]: # (Optional But higlhy recommended Specify *NONE* if missing)
 [//]: # (#### Relevant Documentation:)
 
+### Fix on legacy archetype failing with vro:pull (when workflow folder path name contains special characters(&))
 
-## Upgrade procedure
-[//]: # (Explain in details if something needs to be done)
+#### Previous Behaviour
 
-[//]: # (## Changelog:)
-[//]: # (Pull request links)
+When executing a vro:pull command on a legacy archetype, the command will fail if the workflow path names contain
+special characters such as '&'.
+
+When executing a vro:pull command on JavaScript archetype, it succeeds, ignoring the
+special characters in the folder name.
+
+#### Current Behaviour
+
+When executing a vro:pull command on either Legacy or JS archetype, if the workflows paths contains special character(&),
+the command will successfully execute, ignoring folders with special characters(&).
+
+/test1/test&/test2 -> /test1/test2
+
+Additionally, a warning message will be printed on the console highlighting the issue with unsupported characters.
 
 ### Fix SSH Session methods type
 
@@ -52,3 +64,10 @@ When using SSH with typescript, the `error` and `state` methods has the type `vo
 #### Current Behavior
 
 Method `error` and `state` should return type `String` instead of type `void`
+
+
+## Upgrade procedure
+[//]: # (Explain in details if something needs to be done)
+
+[//]: # (## Changelog:)
+[//]: # (Pull request links)

--- a/docs/versions/latest/Release.md
+++ b/docs/versions/latest/Release.md
@@ -40,16 +40,13 @@
 
 #### Previous Behaviour
 
-When executing a vro:pull command on a legacy archetype, the command will fail if the workflow path names contain
-special characters such as '&'.
+When executing a vro:pull command on a legacy archetype, the command will fail if the workflow path names contain special characters such as '&'.
 
-When executing a vro:pull command on JavaScript archetype, it succeeds, ignoring the
-special characters in the folder name.
+When executing a vro:pull command on JavaScript archetype, it succeeds, ignoring the special characters in the folder name.
 
 #### Current Behaviour
 
-When executing a vro:pull command on either Legacy or JS archetype, if the workflows paths contains special character(&),
-the command will successfully execute, ignoring folders with special characters(&).
+When executing a vro:pull command on either Legacy or JS archetype, if the workflows paths contains special character(&), the command will successfully execute, ignoring folders with special characters(&).
 
 /test1/test&/test2 -> /test1/test2
 

--- a/docs/versions/latest/Release.md
+++ b/docs/versions/latest/Release.md
@@ -40,17 +40,11 @@
 
 #### Previous Behaviour
 
-When executing a vro:pull command on a legacy archetype, the command will fail if the workflow path names contain special characters such as '&'.
-
-When executing a vro:pull command on JavaScript archetype, it succeeds, ignoring the special characters in the folder name.
+When executing a vro:pull command on a legacy archetype, the command will fail without proper error if the workflow paths contains special characters such as '&'.
 
 #### Current Behaviour
 
-When executing a vro:pull command on either Legacy or JS archetype, if the workflows paths contains special character(&), the command will successfully execute, ignoring folders with special characters(&).
-
-/test1/test&/test2 -> /test1/test2
-
-Additionally, a warning message will be printed on the console highlighting the issue with unsupported characters.
+When executing a vro:pull command on a legacy archetype, if the workflows paths contains special character(&), the command will fail but will provide descriptive error message.
 
 ### Fix SSH Session methods type
 

--- a/typescript/vropkg/src/parse/flat.ts
+++ b/typescript/vropkg/src/parse/flat.ts
@@ -19,7 +19,7 @@ import * as path from "path";
 import * as winston from "winston";
 import * as a from "../packaging";
 import * as t from "../types";
-import { read, xml, xmlGet, xmlToCategory, xmlToTag, xmlToAction, xmlChildNamed, getCommentFromJavadoc, getWorkflowItems } from "./util";
+import { read, xml, xmlGet, xmlToCategory, xmlToTag, xmlToAction, xmlChildNamed, getCommentFromJavadoc, getWorkflowItems, validateWorkflowPath} from "./util";
 import { exist } from "../util";
 import { WINSTON_CONFIGURATION, FORM_ITEM_TEMPLATE, WORKFLOW_ITEM_INPUT_TYPE, DEFAULT_FORM_NAME, DEFAULT_FORM_FILE_NAME, VSO_RESOURCE_INF } from "../constants";
 
@@ -37,6 +37,7 @@ const parseFlatElement = async (elementInfoPath: string): Promise<t.VroNativeEle
 
     let infoXml = xml(read(elementInfoPath));
     let categoriesXml = xml(read(elementCategoryPath));
+    validateWorkflowPath(categoriesXml);
     let categoryPath = xmlToCategory(categoriesXml);
     let description = xmlGet(infoXml, "description");
     // input forms (applicable for vRO workflows only)

--- a/typescript/vropkg/src/parse/util.ts
+++ b/typescript/vropkg/src/parse/util.ts
@@ -32,9 +32,9 @@ export const stringToCategory = (category): string[] => category.split(/(?<!\/)\
     .map(path => { return path.replace(/\//g, "") })
     .filter(path => path != "");
 export const validateWorkflowPath = (xmlDocument) => {
-	const invalidElement = xmlDocument.children.find((e) => e.attr.name == null || e.attr.name =='undefined');
-	if (invalidElement && invalidElement.firstChild!=null){
-		console.warn(`Unsupported characters detected in the workflow path. The value "${invalidElement.firstChild.val}" contains characters that are not allowed. The folders with special characters will be ignored.`);
+	const invalidElement = xmlDocument.children.find((e) => e.attr.name === null || e.attr.name === undefined);
+	if (invalidElement && invalidElement.firstChild !== null){
+		throw new Error(`Unsupported characters detected in the workflow path. The value "${invalidElement.firstChild.val}" contains characters that are not allowed.`);
 	}
 };
 export const xmlToAction = (file: string, bundlePath: string, name: string, comment: string, description: string, tags: Array<string>): t.VroActionData => {

--- a/typescript/vropkg/src/parse/util.ts
+++ b/typescript/vropkg/src/parse/util.ts
@@ -31,7 +31,12 @@ export const xmlToTag = (tags): string[] => tags.children.filter(e => e.attr && 
 export const stringToCategory = (category): string[] => category.split(/(?<!\/)\./)
     .map(path => { return path.replace(/\//g, "") })
     .filter(path => path != "");
-
+export const validateWorkflowPath = (xmlDocument) => {
+	const invalidElement = xmlDocument.children.find((e) => e.attr.name == null || e.attr.name =='undefined');
+	if (invalidElement && invalidElement.firstChild!=null){
+		console.warn(`Unsupported characters detected in the workflow path. The value "${invalidElement.firstChild.val}" contains characters that are not allowed. The folders with special characters will be ignored.`);
+	}
+};
 export const xmlToAction = (file: string, bundlePath: string, name: string, comment: string, description: string, tags: Array<string>): t.VroActionData => {
     let type = "" + t.VroElementType.ScriptModule;
     if (!exist(file)) {

--- a/typescript/vropkg/src/serialize/tree.ts
+++ b/typescript/vropkg/src/serialize/tree.ts
@@ -100,9 +100,6 @@ const serializeTreeElement = async (context: any, element: t.VroNativeElement): 
     } else {
         xInfo.ele("comment").cdata(complexActionComment(element));
     }
-    if (element?.categoryPath) {
-		element.categoryPath = element?.categoryPath.filter(x => x!=null);
-	}
     const categoryPathKey = element.type == t.VroElementType.ScriptModule
         ? element?.categoryPath
         : element?.categoryPath.map(c => c.replace(/\./g, "/."))

--- a/typescript/vropkg/src/serialize/tree.ts
+++ b/typescript/vropkg/src/serialize/tree.ts
@@ -100,7 +100,9 @@ const serializeTreeElement = async (context: any, element: t.VroNativeElement): 
     } else {
         xInfo.ele("comment").cdata(complexActionComment(element));
     }
-
+    if (element?.categoryPath) {
+		element.categoryPath = element?.categoryPath.filter(x => x!=null);
+	}
     const categoryPathKey = element.type == t.VroElementType.ScriptModule
         ? element?.categoryPath
         : element?.categoryPath.map(c => c.replace(/\./g, "/."))

